core/src/main/java/org/apache/brooklyn/core/typereg/BasicManagedBundle.java

```diff
@@ -0,0 +1,175 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.typereg;
+
+import java.io.File;
+import java.util.Map;
+
+import org.apache.brooklyn.api.mgmt.rebind.RebindSupport;
+import org.apache.brooklyn.api.typereg.ManagedBundle;
+import org.apache.brooklyn.api.typereg.OsgiBundleWithUrl;
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.mgmt.rebind.BasicManagedBundleRebindSupport;
+import org.apache.brooklyn.core.objs.AbstractBrooklynObject;
+import org.apache.brooklyn.core.objs.BrooklynObjectInternal;
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+
+public class BasicManagedBundle extends AbstractBrooklynObject implements ManagedBundle, BrooklynObjectInternal {
+
+    private String symbolicName;
+    private String version;
+    private String url;
+    private transient File localFileWhenJustUploaded;
+
+    /** Creates an empty one, with an ID, expecting other fields will be populated. */
+    public BasicManagedBundle() {}
+
+    public BasicManagedBundle(String name, String version, String url) {
+        if (name == null && version == null) {
+            Preconditions.checkNotNull(url, "Either a URL or both name and version are required");
+        } else {
+            Preconditions.checkNotNull(name, "Either a URL or both name and version are required");
+            Preconditions.checkNotNull(version, "Either a URL or both name and version are required");
+        }
+
+        this.symbolicName = name;
+        this.version = version;
+        this.url = url;
+    }
+    
+    @Override
+    public boolean isNameResolved() {
+        return symbolicName != null && version != null;
+    }
+    
+    @Override
+    public String getSymbolicName() {
+        return symbolicName;
+    }
+
+    public void setSymbolicName(String symbolicName) {
+        this.symbolicName = symbolicName;
+    }
+    
+    @Override
+    public String getVersion() {
+        return version;
+    }
+
+    public void setVersion(String version) {
+        this.version = version;
+    }
+    
+    @Override
+    public String getUrl() {
+        return url;
+    }
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
+
+    /** This is cached on the object when just uploaded, then deleted after it has been persisted. */
+    @Beta
+    public void setTempLocalFileWhenJustUploaded(File localFileWhenJustUploaded) {
+        this.localFileWhenJustUploaded = localFileWhenJustUploaded;
+    }
+    @Beta
+    public File getTempLocalFileWhenJustUploaded() {
+        return localFileWhenJustUploaded;
+    }
+    
+    @Override
+    public String getOsgiUniqueUrl() {
+        return "brooklyn:"+getId();
```

Comment:
Would it be worth also encoding the name and version in here so that  the URL is more meaningful, at least to a person who should ever happen to read it, even if it's not strictly needed from Karaf's point of view?